test/e2e/validators.go

```diff
@@ -72,6 +73,17 @@ func clientGet(key types.NamespacedName, obj runtime.Object) error {
 	})
 }
 
+func isRetryableError(err error) bool {
```

Comment:
@SaschaSchwarze0 i have a concern in here, you keep adding more selective errors, it will very simple if you just retry for any error that happens. I know you have a concern about this, but maybe is time to fix that concern in the test implementation? if I recall correctly the concern was around:

> I will get errors when retrieving resources because they are not longer there, therefore I should only query for particular errors

I think the above should not be happening, if a for example a buildrun is deleted, then nobody should be querying for that. I can help you on this one if needed.

Reply:
@qu1queee, my general opinion is that it is NOT correct to retry any error that happens on a call to a server. Instead, the majority of errors can be classified clearly into retryable (like 503) or not retryable (like 400) while other errors are fuzzy. 500 is such a fuzzy thing because it means that a programming error on the server happened. This is the case for the last two errors of the `if` conditions. Whether it is worth retrying or not, I here try to detect based on the error message. The alternative would be to retry all 500s (`apierrors.IsInternalError(err)`). I feel a little uncomfortable on that as this potentially hides race conditions that can be fixed (though in this project, the server side of the kube client is no code that comes from this project).

Regarding your specific concern about the 404, I think, it is perfectly valid that a test case verifies whether a resource still exists or not. There are two options to do this: (1) do a get by primary key and expect no error or 404, (2) do a query on the primary key and expect the result size to be 1 or 0. I have no personal preference on which option to use. Generally retrying 404s in a test case I feel not comfortable. I could create resource A and expect resource B then also to be available immediately.

Reply:
All the errors are mainly server calls that failed. We cannot continue catching errors later that then we need to add into this list of different errors types. We need to stabilize this and if then we continue having red PRs, then we need to have a different discussion.

 The whole idea of polling during a period of time is to retry independently of the error type, and give more time to the system to react properly. Also, we are not hiding anything, if there is an internal issue it will fail eventually. Plus our travis is not stable.

>  I think, it is perfectly valid that a test case verifies whether a resource still exists or not

yes, but this will be a single call, it does not require to be polling.

Reply:
Hi @qu1queee, I am not completely getting your complaint.

> All the errors are mainly server calls that failed. We cannot continue catching errors later that then we need to add into this list of different errors types. We need to stabilize this and if then we continue having red PRs, then we need to have a different discussion.

I agree that it would be best if we could assess upfront the full list of potential errors and add those to the retry list that are suitable for retry. And if there is another IsSomething function in https://godoc.org/k8s.io/apimachinery/pkg/api/errors that clearly indicates this, I am happy to add it to the list. But as explained, the error we are addressing in this PR is a 500. Root cause connection problem in IAM (happens in IKS, not in Travis here). And for 500 I am hesitant in retrying them all to not hide potentially fixable issues.

> Also, we are not hiding anything, if there is an internal issue it will fail eventually.

I agree that any programming error in code that fails always will eventually also fail when retried n times. But code that only fails in 5 or 10 % of the cases due to a race condition, will be hidden by retries.

-- 

If I remember correctly, in another PR related to this code I already mentioned that I do not like it and mean it does not look nice. Honestly, if I remember previous projects, retry logic never looked nice. One relies on Exception types (Java world) and status codes and sometimes even message strings to determine whether an error is retriable or not. But, conceptually, I think that in an architecture with different independently running (micro)services communicating over network, one requires to implement the necessary mechanisms to live with temporary failures like the ones in my `if` clause = those errors that are retriable. One should - on the other hand - also not overload services with requests causing non-retriable errors like 400 bad requests. That's why I consider unreasonable retries as a bad practice.

Reply:
I was on the wrong line of though. This PR is ok for now. But let me layout how the "polling" mechanism in the test should be in my opinion:

Ideally I wanted to only avoid one single error, basically "if apierrors.IsNotFound(err)" return "false,nil" to continue polling. The rest of the errors should be giving you back a "false,error", meaning you stop polling and fail the test. This is basically bailing out as soon as an error is not the one related to "not found" because you only care about if a resource is available. I think I never fully agree on continue polling although some other errors happened, I think i agreed on this because travis seems to be overloaded with our amount of parallel tests. 

Because of the travis situation and us not having another way to tackle this, Im fine with adding more selective error checks that can be ignored. But based on how I see the polling, I would like to change this implementation in the future.

Also, I´m not sure if we should be mixing IKS issues with the way we do in on this travis, where IKS/IAM is totally not related. It will be hard for any other developer to understand why we need this PR, if they have no context on IKS/IAM and how it relates to a 500 http code. If there is no issue from our side to this IKS error, it would be good to have this and find the root cause. While this will have an impact on customers when triggering builds.